parallelize sort

Ryo-not-rio opened the pull request:
- use __gnu_parallel::sort for gcc compilations
- add a parallelized version of std::sort and std::stable_sort for non gcc compilations

Using __gnu_parallel::sort:
provides ~3.7x speed up for length 50000 sorts with NUM_THREADS=16 and NUM_THREADS=4 on aarch64

The performance is measured using the following script:
```python
import torch
import torch.autograd.profiler as profiler

torch.manual_seed(0)

N = 50000
x = torch.randn(N, dtype=torch.float)

with profiler.profile(with_stack=True, profile_memory=False, record_shapes=True) as prof:
    for i in range(1000):
        _, _ = torch.sort(x)

print(prof.key_averages(group_by_input_shape=True).table(sort_by='self_cpu_time_total', row_limit=10))

```


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @malfet @snadampal @milpuz01

Skylion007 commented:
std C++17 has a parallel sort algorithm builtin. Any reason we aren't using that? Because our supported compilers do not have full C++17 support?

nikitaved commented on aten/src/ATen/native/cpu/SortingKernel.cpp:
```diff
@@ -139,30 +143,82 @@ static void parallel_sort1d_kernel(
     if (stable) {
-      std::stable_sort(composite_accessor, composite_accessor + dim_size,
-        KeyValueCompDesc<scalar_t>());
+      parallel_sort(composite_accessor, composite_accessor + dim_size,
+            KeyValueCompDesc<scalar_t>(), std::stable_sort<CompositeIt, KeyValueCompDesc<scalar_t>>);
```
Since PyTorch now lives in c++17, can't we just use `std::sort` with some `std::execution::parallel_*_policy`?

nikitaved replied:
Sorry, @Skylion007 , stepped on your feet :)

Ryo-not-rio replied:
Hi I've tried this locally both within PyTorch and with a minimal reproducible example using C++17 but it seems to have no effect, which is why we've opted for __gnu_parallel::sort

nikitaved replied:
It looks like there might be a soft dependence on TBB https://stackoverflow.com/questions/51031060/are-c17-parallel-algorithms-implemented-already.

Ryo-not-rio replied:
Hi, thanks for the review. __gnu_prallel seems to be about 2x faster so I've put it as a default option for gcc but I've changed the backup option to the suggested method using std::execution

Skylion007 replied:
This is surprising to me. Is it because the version of GCC is too old to have an implementation for the execution policy (ie falling back to serial implementation), or that there is some stdlib edge cases / overhead that is slowing things down?

nikitaved replied:
https://gcc.gnu.org/onlinedocs/libstdc++/manual/parallel_mode_using.html -- I actually cannot grep for `D_GLIBCXX_PARALLEL`.

Ryo-not-rio replied:
@Skylion007  there must be a different implementation or an overhead using stdlib because using the execution policy still provides a benefit, just not as much as __gnu_parallel

Ryo-not-rio replied:
@nikitaved My understanding was that if #include <parallel/algorithm>, we wouldn't need to add `D_GLIBCXX_PARALLEL`

nikitaved replied:
@Ryo-not-rio, yes, but it is only used for the GNU extensions, no?

Ryo-not-rio replied:
I'm a little hesitant in enabling parallelization for everything at once since this may potentially introduce bugs to unforseen areas

nikitaved commented on aten/src/ATen/native/cpu/SortingKernel.cpp:
```diff
@@ -139,30 +143,82 @@ static void parallel_sort1d_kernel(
+
+  if (num_threads > 1) {
+    const int64_t merge_threads = (num_threads - ((int) chunk_size * num_threads != n)) / 2;
+    at::parallel_for(
+      0, chunk_size * 2 * merge_threads, chunk_size * 2, [&](int64_t begin, int64_t end) {
+        std::inplace_merge(first + begin, first + begin + chunk_size, first + end, comp);
+      }
+    );
+
+    for (int i=0; i < merge_threads - 1; i++) {
+      std::inplace_merge(first, first + chunk_size * 2 * (i + 1), first + chunk_size * 2 * (i + 2), comp);
+    }
```
Merge can be done in parallel and in a hierarchical manner. Here it looks like the parallelization is done only on a single level.

nikitaved replied:
Actually, sounds like a good thing to have at some point -- a generalized parallel implementation of divide-and-conquer types of algorithms...

Ryo-not-rio replied:
yes, although the conquer part is quite specific to each algorithm

Ryo-not-rio committed:
parallelize sort

- use __gnu_parallel::sort for gcc compilations
- add a parallelized version of std::sort and std::stable_sort for non gcc compilations

Using __gnu_parallel::sort:
provides ~3.7x speed up for length 50000 sorts with NUM_THREADS=16 and NUM_THREADS=4 on aarch64

Otherwise:
provides ~2x speed up for length 50000 sorts with NUM_THREADS=16 and NUM_THREADS=4 on aarch64

Ryo-not-rio commented on CMakeLists.txt:
```diff
@@ -538,8 +538,7 @@ option(BUILD_EXECUTORCH "Master flag to build Executorch" ON)
   set(CMAKE_SHARED_LINKER_FLAGS
-      "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--no-as-needed")
-
+      "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--no-as-needed -ltbb")
```
Done!

ng-05 commented:
@pytorchbot label "module: arm"

malfet commented on .ci/docker/manywheel/Dockerfile_2_28_aarch64:
```diff
@@ -28,6 +28,8 @@ RUN yum install -y \
   openssl-devel \
   patch \
   perl \
+  tbb \
```
Why do you need tbb here?

malfet reviewed (changes requested):
Is paralle sort algorithm a GNU C++ extension or something else? And why do you need tbb here.

It feels wrong adding random compiler extensions which are yet not planned to be included into next compiler standard.

If this is something that is available via say C++20 standard, I would be fine with adding tentative standard flag, but if enabling this requires new runtime dependency, than it's probably a no-starter to be included in any source builds

Ryo-not-rio commented:
> Is paralle sort algorithm a GNU C++ extension or something else? And why do you need tbb here.
> 
> It feels wrong adding random compiler extensions which are yet not planned to be included into next compiler standard.
> 
> If this is something that is available via say C++20 standard, I would be fine with adding tentative standard flag, but if enabling this requires new runtime dependency, than it's probably a no-starter to be included in any source builds

This is a libstdc++ extension listed in https://gcc.gnu.org/onlinedocs/libstdc++/manual/parallel_mode.html. `tbb` seems to be a requirement for the actual parallelism to work as we did not see any benefit without it. We also tried the execution policy for std::sort but did not see any improvements

Skylion007 commented:
More likely the issue is that parllallel policy for the stdlib were some of the last C++17 features to be implemented so I am not even sure if our stdlib version actually supports them (vs falling back on the serial implementation)

Ryo-not-rio committed:
Remove tbb and execution policy from parallel sort

Ryo-not-rio commented:
@malfet @ng-05 I have removed the execution policy fallback as that is what was requiring tbb

malfet commented:
> provides ~2x speed up for length 50000 sorts with NUM_THREADS=16 and NUM_THREADS=4 on aarch64

@Ryo-not-rio can you please share the script in PR description that could be used to reproduce those results? (Though I guess it's a comment from TBB case, as I don't see how it makes any difference unless GCC extension is used) Which again, makes me wonder, if this extension is slated for C++23 or is it always an extension?

annop-w commented:
@malfet The parallel mode extension has been around for quite some time already. Please see the documentation https://gcc.gnu.org/onlinedocs/libstdc++/manual/parallel_mode.html

Ryo-not-rio committed:
Use -D_GLIBCXX_PARALLEL to parallelize sort

Ryo-not-rio commented:
@nikitaved @malfet @annop-w I've updated the code so we're relying on the `-D_GLIBCXX_PARALLEL` flag to parallelize the sort function. I've limited the flag to just be applied on SortingKernel.cpp to prevent unintended consequences elsewhere

Ryo-not-rio commented:
@pytorchbot label "arm priority"

malfet commented on cmake/Codegen.cmake:
```diff
@@ -383,6 +383,16 @@ if(INTERN_BUILD_ATEN_OPS)
+      # Only parallelize the SortingKernel for now to avoid side effects
+      if (${NAME} STREQUAL "native/cpu/SortingKernel.cpp")
+        if(MSVC)
+            set(EXTRA_FLAGS "${EXTRA_FLAGS} /D_GLIBCXX_PARALLEL")
```
How was that tested? And GNU libstdc++ aren't really used on Windows by any of the builds, is it? (And specifically not by VisualC++ compiler)

Ryo-not-rio replied:
Removed

malfet commented on cmake/Codegen.cmake:
```diff
@@ -383,6 +383,16 @@ if(INTERN_BUILD_ATEN_OPS)
+        if(MSVC)
+            set(EXTRA_FLAGS "${EXTRA_FLAGS} /D_GLIBCXX_PARALLEL")
+        else(MSVC)
+            set(EXTRA_FLAGS "${EXTRA_FLAGS} -D_GLIBCXX_PARALLEL")
```
```suggestion
            string(APPEND EXTRA_FLAGS " -D_GLIBCXX_PARALLEL")
```

Ryo-not-rio replied:
Done

malfet reviewed (commented):
Changes to SortingKernel.cpp looks like a purely formatting ones, can you submit them as separate PR?
CMake change looks OK, minus MSVC changes (which to the best of my knowledge will not work)

Ryo-not-rio commented:
@malfet Thank you for the review, I have addressed your comments

malfet commented:
@pytorchbot merge -r

malfet commented on cmake/Codegen.cmake:
```diff
@@ -383,6 +383,12 @@ if(INTERN_BUILD_ATEN_OPS)
       endif(MSVC)
+
+      # Only parallelize the SortingKernel for now to avoid side effects
+      if (${NAME} STREQUAL "native/cpu/SortingKernel.cpp" AND NOT MSVC)
```
```suggestion
      if (${NAME} STREQUAL "native/cpu/SortingKernel.cpp" AND NOT MSVC AND USE_OPENMP)
```

annop-w commented:
@pytorchbot merge

Ryo-not-rio commented:
@pytorchbot merge

malfet commented:
@pytorchbot merge -r

Ryo-not-rio commented:
@pytorchbot merge -r

malfet commented:
@pytorchbot merge -i
